Search wheels for `.dist-info` directories

stefanor opened the pull request:
Some wheels don't use normalized names for their `.dist-info` directories, so search the wheel for them.

Fixes: #134

stefanor commented on src/installer/sources.py:
```diff
@@ -122,6 +122,33 @@ def open(cls, path: "os.PathLike[str]") -> Iterator["WheelFile"]:
+            norm_file_dname = normalize_distribution_name(self.distribution)
+            assert (
+                norm_di_dname == norm_file_dname
+            ), "Wheel .dist-info directory doesn't match wheel filename"
```
This isn't exactly the same check that pip does. It knows what distribution name it's expecting to install, so that's what it's looking for.

In the wheels I've seen that we hit this bug on, this assertion would be true.

pradyunsg replied:
We also know the name, via `self.distribution`.

pradyunsg replied:
Ah, right, in case of a mismatch. Nvm me.

ArmaanT commented:
Is there anything I can do to help get this merged? I've been running into this problem when using bazel to install wheels with capitalization issues.

stefanor commented:
Ping?

pradyunsg commented on src/installer/utils.py:
```diff
@@ -93,6 +93,14 @@ def parse_metadata_file(contents: str) -> Message:
     return feed_parser.close()
 
 
+def normalize_distribution_name(name: str) -> str:
```
Hmm... this might be easier to parse for folks if this has the same name as the `packaging` function.

```suggestion
def canonicalize_name(name: str) -> str:
```

pradyunsg commented on src/installer/sources.py:
```diff
@@ -122,6 +122,33 @@ def open(cls, path: "os.PathLike[str]") -> Iterator["WheelFile"]:
+
+            # NAME-VER.dist-info
+            di_dname = dist_info_dir.rsplit("-", 2)[0]
+            norm_di_dname = normalize_distribution_name(di_dname)
+            norm_file_dname = normalize_distribution_name(self.distribution)
```
```suggestion
            norm_di_dname = canonicalize_name(di_dname)
            norm_file_dname = canonicalize_name(self.distribution)
```

pradyunsg reviewed (dismissed):
LGTM, with one non-blocking suggestion.

Lemme know if you'd prefer that I merge without that change. :)

pradyunsg commented:
TestWheelFile.test_finds_dist_info is failing on 3.7.

stefanor commented:
Applied your name suggestion, and rebased on top of master.

> TestWheelFile.test_finds_dist_info is failing on 3.7.

Added a workaround for that in a separate commit, that's easy to revert later.

pradyunsg commented:
Thanks @stefanor!

PS: I've filed #157 for a few follow up changes that I'd like to make. :)
